File: src/draft/replacement.py
"""Replacement level for a specific league shape.

Replacement level is the whole ballgame in a 14-team league: it is what converts
"projected points" (which says QBs are the most valuable players in football)
into "value over replacement" (which says they are not, because the 14th QB is
nearly as good as the 3rd).

We derive it rather than hardcoding it. Base starters are known from the league
config; the FLEX slots are allocated to whichever RB/WR/TE actually deserve them
given the projections, so replacement level responds to the shape of this year's
player pool instead of a rule of thumb.
"""

from __future__ import annotations

import polars as pl

from src.config import LEAGUE


def allocate_flex(proj: pl.DataFrame, league: dict | None = None) -> dict[str, int]:
    """Return the number of *starters* at each position, flex included.

    Base starters are filled first. The remaining flex slots go to the best
    players left across all flex-eligible positions, which is what managers
    collectively do.
    """
    league = league or LEAGUE
    teams = league["teams"]
    starters = league["starters"]
    eligible = league["flex_eligible"]

    counts = {pos: starters.get(pos, 0) * teams for pos in proj["pos"].unique()}

    flex_slots = starters.get("FLEX", 0) * teams
    if flex_slots:
        # Everyone at a flex-eligible position who is not already a base starter,
        # ranked by projection; the top `flex_slots` of them take the flex spots.
        pool = (
            proj.filter(pl.col("pos").is_in(eligible))
            .with_columns(
                pl.col("proj_points").rank("ordinal", descending=True).over("pos").alias("r")
            )
            .filter(pl.col("r") > pl.col("pos").replace_strict(counts, default=0))
            .sort("proj_points", descending=True)
            .head(flex_slots)
        )
        for pos, n in pool.group_by("pos").len().iter_rows():
            counts[pos] = counts.get(pos, 0) + n

    return counts


def replacement_levels(proj: pl.DataFrame, league: dict | None = None) -> dict[str, float]:
    """Projected points of the best non-starter at each position."""
    counts = allocate_flex(proj, league)
    levels: dict[str, float] = {}

    for pos in proj["pos"].unique().to_list():
        ranked = proj.filter(pl.col("pos") == pos).sort("proj_points", descending=True)
        idx = counts.get(pos, 0)  # 0-based index of the first non-starter
        if ranked.height == 0:
            continue
        if idx >= ranked.height:
            # Pool is shallower than the number of starting slots: replacement is
            # the worst rostered player, i.e. effectively nothing is free here.
            levels[pos] = float(ranked["proj_points"][-1])
        else:
            levels[pos] = float(ranked["proj_points"][idx])
    return levels


def add_vor(proj: pl.DataFrame, league: dict | None = None) -> pl.DataFrame:
    """Attach value over replacement, and rank the board by it."""
    levels = replacement_levels(proj, league)
    return (
        proj.with_columns(
            pl.col("pos").replace_strict(levels, default=0.0).alias("replacement")
        )
        .with_columns((pl.col("proj_points") - pl.col("replacement")).alias("vor"))
        .with_columns(pl.col("vor").rank("ordinal", descending=True).alias("vor_rank"))
        .sort("vor", descending=True)
    )

File: v2/lib/replacement.ts
/**
 * Replacement level for a specific league shape.
 *
 * Port of `src/draft/replacement.py`. Replacement level is what converts
 * "projected points" (which says quarterbacks are the most valuable players in
 * football) into "value over replacement" (which says they are not, because in
 * a 14-team league the 14th quarterback is nearly as good as the 3rd).
 *
 * It is derived, not hardcoded. Base starters come from the league config; the
 * FLEX slots go to whichever flex-eligible players actually earn them given the
 * projections, so replacement responds to the shape of this year's pool rather
 * than a rule of thumb.
 */

import type { LeagueConfig, Player, Position } from "./types";

/** Positions eligible for a SUPERFLEX slot: the flex positions plus QB. */
function superflexEligible(league: LeagueConfig): Position[] {
  const set = new Set<Position>(league.flexEligible);
  set.add("QB");
  return [...set];
}

/**
 * Number of *starters* at each position across the league, flex included.
 *
 * Base starters are filled first. Remaining flex slots go to the best players
 * left across all flex-eligible positions, which is what managers collectively
 * do.
 */
export function allocateFlex(
  players: Player[],
  league: LeagueConfig,
): Record<string, number> {
  const teams = league.teams;
  const counts: Record<string, number> = {};
  for (const p of players) {
    if (!(p.pos in counts)) counts[p.pos] = (league.starters[p.pos] ?? 0) * teams;
  }

  // Rank within position by projection, descending, 1-based.
  const byPos = new Map<Position, Player[]>();
  for (const p of players) {
    if (!byPos.has(p.pos)) byPos.set(p.pos, []);
    byPos.get(p.pos)!.push(p);
  }
  for (const list of byPos.values()) list.sort((a, b) => b.proj_points - a.proj_points);

  const fill = (slotName: "FLEX" | "SUPERFLEX", eligible: Position[]) => {
    const slots = (league.starters[slotName] ?? 0) * teams;
    if (slots <= 0) return;

    // Everyone at an eligible position who is not already a starter.
    const pool: Player[] = [];
    for (const pos of eligible) {
      const list = byPos.get(pos) ?? [];
      pool.push(...list.slice(counts[pos] ?? 0));
    }
    pool.sort((a, b) => b.proj_points - a.proj_points);
    for (const p of pool.slice(0, slots)) counts[p.pos] = (counts[p.pos] ?? 0) + 1;
  };

  fill("FLEX", league.flexEligible);
  fill("SUPERFLEX", superflexEligible(league));

  return counts;
}

/** Projected points of the best non-starter at each position. */
export function replacementLevels(
  players: Player[],
  league: LeagueConfig,
): Record<string, number> {
  const counts = allocateFlex(players, league);
  const levels: Record<string, number> = {};

  const byPos = new Map<Position, Player[]>();
  for (const p of players) {
    if (!byPos.has(p.pos)) byPos.set(p.pos, []);
    byPos.get(p.pos)!.push(p);
  }

  for (const [pos, list] of byPos) {
    if (list.length === 0) continue;
    const ranked = [...list].sort((a, b) => b.proj_points - a.proj_points);
    const idx = counts[pos] ?? 0;
    levels[pos] =
      idx >= ranked.length
        // Pool shallower than the number of starting slots: replacement is the
        // worst rostered player, i.e. effectively nothing here is free.
        ? ranked[ranked.length - 1].proj_points
        : ranked[idx].proj_points;
  }
  return levels;
}

/** Attach value over replacement and rank the board by it. */
export function addVor(players: Player[], league: LeagueConfig): Player[] {
  const levels = replacementLevels(players, league);

  const scored = players.map((p) => {
    const replacement = levels[p.pos] ?? 0;
    return { ...p, replacement, vor: p.proj_points - replacement };
  });

  // Ties broken by player id. The Python breaks them by whatever order the
  // frame happened to be in, which is fine for a single-shot script but would
  // let the board reshuffle between renders here. Ties are arbitrary either
  // way; this at least makes them stable.
  scored.sort((a, b) => b.vor - a.vor || (a.player_id < b.player_id ? -1 : 1));
  scored.forEach((p, i) => {
    p.vor_rank = i + 1;
  });
  return scored;
}
